### ISSUE TITLE
needless_update false positive when struct definition or struct expression uses `cfg`

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#13076]
### Summary

There are cases where a struct expression using [functional update syntax](https://doc.rust-lang.org/1.79.0/reference/expressions/struct-expr.html#functional-update-syntax) is the best way to express what I want, but some of the fields in the struct are conditionally compiled, and under some feature combinations it can be possible that the `..` part of the struct expression contains no fields.

Example: some of the method implementations in https://github.com/dtolnay/syn/blob/2.0.70/src/fixup.rs.

### Lint Name

needless_update

### Reproducer

Struct definition uses conditional compilation:

```rust
struct Struct {
    #[cfg(feature = "a")]
    a: bool,

    #[cfg(feature = "b")]
    b: bool,
}

impl Struct {
    #[cfg(feature = "a")]
    fn a(self) -> Self {
        Struct { a: true, ..self }
    }
}
```

```console
$ cargo clippy --features a
warning: struct update has no effect, all the fields in the struct have already been specified
  --> src/main.rs:12:29
   |
12 |         Struct { a: true, ..self }
   |                             ^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_update
   = note: `#[warn(clippy::needless_update)]` on by default
```

---

Struct expression uses conditional compilation:

```rust
struct Struct {
    a: bool,
}

impl Struct {
    fn a(self) -> Self {
        Struct {
            #[cfg(feature = "a")]
            a: true,
            ..self
        }
    }
}
```

```console
$ cargo clippy --features a
warning: struct update has no effect, all the fields in the struct have already been specified
  --> src/main.rs:10:15
   |
10 |             ..self
   |               ^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_update
   = note: `#[warn(clippy::needless_update)]` on by default
```

### Version

```text
rustc 1.81.0-nightly (20ae37c18 2024-07-07)
binary: rustc
commit-hash: 20ae37c18df95f9246c019b04957d23b4164bf7a
commit-date: 2024-07-07
host: x86_64-unknown-linux-gnu
release: 1.81.0-nightly
LLVM version: 18.1.7
```

**Comment by dtolnay** [comment]
One might expect that this can be worked around by using another `cfg` to only include the `..` when it is nonempty:

- `Struct { a: true, #[cfg(feature = "b")] ..self }` (in the first case)
- `Struct { #[cfg(feature = "a")] a: true, #[cfg(not(feature = "a"))] ..self }` (in the second case)

but that is not valid Rust syntax.
